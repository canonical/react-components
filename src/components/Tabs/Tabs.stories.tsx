import React from "react";
import { Meta, StoryObj } from "@storybook/react";

import Row from "../Row";
import Tabs from "./Tabs";

const meta: Meta<typeof Tabs> = {
  component: Tabs,
  tags: ["autodocs"],
};

export default meta;

type Story = StoryObj<typeof Tabs>;

export const Default: Story = {
  name: "Default",

  args: {
    links: [
      {
        active: true,
        label: "Summary",
      },
      {
        active: false,
        label: "Network",
      },
      {
        active: false,
        label: "Storage",
      },
      {
        active: false,
        label: "Settings",
      },
    ],
  },
};

/**
 * To horizontally align the Tabs with other content, they can be contained within a Row component to
provide correct gutters.
 */
export const HorizontallyAligned: Story = {
  render: () => (
    <Row>
      <Tabs
        links={[
          {
            active: true,
            label: "Summary",
          },
          {
            active: false,
            label: "Network",
          },
          {
            active: false,
            label: "Storage",
          },
          {
            active: false,
            label: "Settings",
          },
        ]}
      />
      <p>There should be gutters and the text should be horizontally aligned</p>
    </Row>
  ),

  name: "Horizontally aligned",
};
