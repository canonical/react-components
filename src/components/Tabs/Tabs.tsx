import classNames from "classnames";
import React from "react";
import type { HTMLProps, ElementType, ReactNode, ComponentType } from "react";

import type { ClassName } from "types";

export type TabLink<P = null> = {
  /**
   * Whether the tab link should have active styling.
   */
  active?: boolean;
  /**
   * Optional classes applied to the link element.
   */
  className?: ClassName;
  /**
   * Optional component to be used instead of the default "a" element.
   */
  component?: ElementType | ComponentType<P>;
  /**
   * Label to be displayed inside the tab link.
   */
  label: ReactNode;
  /**
   * Optional classes applied to the "li" element.
   */
  listItemClassName?: string;
} & (HTMLProps<HTMLElement> | P);

export type Props<P = null> = {
  /**
   * Optional classes applied to the parent "nav" element.
   */
  className?: ClassName;
  /**
   * An array of tab link objects.
   */
  links: TabLink<P>[];
  /**
   * Optional classes applied to the "ul" element.
   */
  listClassName?: string;
};

/**
 * This is the [React](https://reactjs.org/) component for Vanilla [Tabs](https://vanillaframework.io/docs/patterns/tabs).
Tabs organise and allow navigation between groups of content that are related and at the same level
of hierarchy.
 */
const Tabs = <P,>({
  className,
  links,
  listClassName,
}: Props<P>): JSX.Element => {
  return (
    <nav className={classNames("p-tabs", className)}>
      <ul className={classNames("p-tabs__list", listClassName)}>
        {links.map((link, i) => {
          const {
            active,
            className,
            component,
            label,
            listItemClassName,
            ...rest
          } = link;
          const Component = component || "a";
          return (
            <li
              className={classNames("p-tabs__item", listItemClassName)}
              key={i}
            >
              <Component
                aria-selected={active}
                className={classNames("p-tabs__link", className)}
                data-testid={`tab-link-${label}`}
                {...rest}
              >
                {label}
              </Component>
            </li>
          );
        })}
      </ul>
    </nav>
  );
};

export default Tabs;
