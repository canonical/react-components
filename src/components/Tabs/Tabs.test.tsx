import { shallow } from "enzyme";
import React from "react";

import Tabs from "./Tabs";

describe("Tabs", () => {
  it("renders", () => {
    const wrapper = shallow(
      <Tabs
        links={[
          {
            active: true,
            href: "/path1",
            label: "label1",
          },
          {
            active: false,
            href: "/path2",
            label: "label2",
          },
        ]}
      />
    );
    expect(wrapper).toMatchSnapshot();
  });

  it("sets an active item correctly", () => {
    const wrapper = shallow(
      <Tabs
        links={[
          {
            active: true,
            label: "label1",
          },
          {
            active: false,
            label: "label2",
          },
        ]}
      />
    );
    expect(wrapper.find("[aria-selected=true]").length).toBe(1);
  });

  it("can set classNames correctly", () => {
    const wrapper = shallow(
      <Tabs
        className="nav-class"
        listClassName="list-class"
        links={[
          {
            className: "link-class",
            label: "label1",
            listItemClassName: "list-item-class",
          },
        ]}
      />
    );
    expect(wrapper.find("nav").props().className.includes("nav-class")).toBe(
      true
    );
    expect(wrapper.find("ul").props().className.includes("list-class")).toBe(
      true
    );
    expect(
      wrapper.find("li").props().className.includes("list-item-class")
    ).toBe(true);
    expect(wrapper.find("a").props().className.includes("link-class")).toBe(
      true
    );
  });

  it("can use custom elements as links", () => {
    const wrapper = shallow(
      <Tabs
        links={[
          {
            component: "div",
            label: "label1",
            to: "/path",
          },
        ]}
      />
    );
    expect(wrapper.find("a[data-testid='tab-link-label1']").exists()).toBe(
      false
    );
    expect(wrapper.find("div[data-testid='tab-link-label1']").exists()).toBe(
      true
    );
  });

  it("can use custom components as links", () => {
    const TestLink = ({ children, ...props }) => (
      <span {...props}>{children}</span>
    );
    const wrapper = shallow(
      <Tabs
        links={[
          {
            component: TestLink,
            label: "label1",
            to: "/path",
          },
        ]}
      />
    );
    expect(wrapper.find("a[data-testid='tab-link-label1']").exists()).toBe(
      false
    );
    expect(
      wrapper.find("TestLink[data-testid='tab-link-label1']").exists()
    ).toBe(true);
  });
});
