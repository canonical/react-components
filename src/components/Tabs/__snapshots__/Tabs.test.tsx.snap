// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tabs renders 1`] = `
<nav
  className="p-tabs"
>
  <ul
    className="p-tabs__list"
  >
    <li
      className="p-tabs__item"
      key="0"
    >
      <a
        aria-selected={true}
        className="p-tabs__link"
        data-testid="tab-link-label1"
        href="/path1"
      >
        label1
      </a>
    </li>
    <li
      className="p-tabs__item"
      key="1"
    >
      <a
        aria-selected={false}
        className="p-tabs__link"
        data-testid="tab-link-label2"
        href="/path2"
      >
        label2
      </a>
    </li>
  </ul>
</nav>
`;
